File: Contest/Weekly Contest 311/4th Problem/JAVA Code/2416. Sum of Prefix Scores of Strings.java
class Solution {
    
    public static class TrieNode{
        TrieNode[] node;
        int count;
        TrieNode() {
            node = new TrieNode[26];
        }
    }
    
    public static void add(TrieNode root,String s) {
        for(int i=0;i<s.length();i++) {
            int idx = s.charAt(i)-'a';
            if(root.node[idx]==null) root.node[idx] = new TrieNode();
            root = root.node[idx];
            root.count++;
        }
    }
    
    public int[] sumPrefixScores(String[] words) {
        TrieNode root = new TrieNode();
        
        for(String word : words) add(root,word);
        
        int[] ans = new int[words.length];
        for(int i=0;i<ans.length;i++) {
            ans[i] = find(words[i],root);
        }
        
        return ans;
    }
    
    public static int find(String s,TrieNode root) {
        int ans=0;
        for(int i=0;i<s.length();i++) {
            int idx = s.charAt(i)-'a';
            root=root.node[idx];
            ans+=root.count;
        }
        return ans;
    }
    
}

File: Contest/Weekly Contest 311/4th Problem/C++ Code/2416. Sum of Prefix Scores of Strings.cpp
class Solution {
public:
    class TrieNode{
        public: TrieNode* node[26];
        public: int count=0;
        public: TrieNode(){
            for(int i=0;i<26;i++) {
                node[i]=NULL;
                count=0;
            }
        }
    };
    
    void add(TrieNode* root,string s) {
        for(int i=0;i<s.length();i++) {
            int idx = s[i]-'a';
            if(root->node[idx]==NULL) root->node[idx] = new TrieNode();
            root = root->node[idx];
            root->count++;
        }
    }
    
    vector<int> sumPrefixScores(vector<string>& words) {
        TrieNode* root = new TrieNode();
        
        for(string word : words) add(root,word);
        
        vector<int> ans;
        for(string word : words) ans.push_back(find(word,root));
        
        return ans;
    }
    
    int find(string s,TrieNode* root) {
        int ans=0;
        for(int i=0;i<s.length();i++) {
            int idx = s[i]-'a';
            root=root->node[idx];
            ans+=root->count;
        }
        return ans;
    }
};
